src/mscorlib/src/System/Runtime/CompilerServices/jithelpers.cs

```diff
@@ -191,5 +191,27 @@ static internal PinningHelper GetPinningHelper(Object o)
             typeof(ArrayPinningHelper).ToString(); // Type used by the actual method body
             throw new InvalidOperationException();
         }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        static internal byte BoolToByteNonNormalized(bool value)
```

Comment:
Why does this return `byte`? `int` would make more sense. A `bool` is 1 byte in size only when stored in memory (e.g. in a class field or array element). It is never 1 byte on the evaluation stack.

Reply:
Maybe this method should be added to `Unsafe`. It basically works around a C# language limitation (like other `Unsafe` methods do) and at the same time it is slightly unsafe due to the possibility of a `bool` not being only 0 or 1.

Reply:
I was hoping that there might be some cases where RyuJIT might elide the `movzx` instruction if it was kept as an 8-bit integral type all the way throughout the call sites. I'm thinking of cases like `myByteArray[idx] = BoolToByte(condition);`, where everything would just compile to a single `mov byte ptr [rdx + rcx], eax` instruction. But I'm not married to returning a 32-bit integer, and if the JIT is smart enough to make this all work anyway then all the better. :)

Reply:
> I'm thinking of cases like myByteArray[idx] = BoolToByte(condition);, where everything would just compile to a single mov byte ptr [rdx + rcx], eax instruction.

I already sent a PR to fix that specific case. The JIT already eliminates `movzx` in some other cases.

At the same time, `movzx` is a cheap instruction (0 latency normally) so it shouldn't have much impact. But if you do see useless `movzx` being generated, file issues. There are a lot of things that the JIT can't do but there are also quite a few things that the JIT could do, provided someone bothers with writing the necessary code.